cpp/include/cudf/detail/utilities/assert.cuh

```diff
@@ -35,3 +35,19 @@
 #else
 #define cudf_assert(e) (static_cast<void>(0))
 #endif
+
+/**
+ * @brief Macro indicating that a location in the code is unreachable.
+ *
+ * This can be used in conjunction with dispatchers to indicate that a function
+ * does not need to return a default value because it has already exhausted all
+ * possible cases in a `switch` statement.
```

Comment:
This makes it sound like it's only relevant for dispatching code. It's fine as an example, but we should make it more clear that it is only one example of many usages. Another common use case would be for type dispatched code paths that will never be executed in device code, e.g., places where we use `cudf_assert` today.

Reply:
For example: https://github.com/rapidsai/cudf/blob/7b9ea2b480ba723e5fc70e17c42a0a93123fef80/cpp/include/cudf/table/row_operators.cuh#L211-L217

Reply:
We should include that `CUDF_UNREACHABLE` should only be used where `CUDF_FAIL` cannot be used whether due to performance or due to being used in device code. In the majority of host code situations we should be throwing an exception in "unreachable" code paths as those usually aren't tight inner loops like they are in device code.

Reply:
I was wondering the same thing. Should we create a macro specifically for this purpose, i.e. `CUDF_FAIL` if on host and `CUDF_UNREACHABLE` if on device?

Reply:
No, I think that would be conflating two different things. There may be plenty of instances where I want to explicitly say some path in host code is unreachable without throwing an exception.

Reply:
Here's a noteworthy case where the need to prefer `CUDF_FAIL` over `CUDF_UNREACHABLE` on the host came up: I noticed a segfault locally when I replaced the type dispatcher's `default` case with `CUDF_UNREACHABLE`. In commit 77c97aeaea29b6efb0bdaba810b8bcbbaba2df54 I realized that we have to call `CUDF_FAIL` on the host, or else [this factory test](https://github.com/rapidsai/cudf/blob/f301b282ee37e31ecdb9b3baa5eb78c38066fd8b/cpp/tests/column/factories_test.cpp#L362-L372) hits a segfault when dispatching to make a non-numerical column from type id `EMPTY`, [one of the tested non-fixed-width types](https://github.com/rapidsai/cudf/blob/c1f20c7be6bc7cab8af1fb20032f05c186d41cfe/cpp/include/cudf_test/type_lists.hpp#L406-L407). Calling `CUDF_FAIL` on the host raises the expected error, while `CUDF_UNREACHABLE` invokes UB and segfaults.

Reply:
Also I updated the docs in ab1cdb9.